_posts/2023-01-08-running-apache-kafka-on-FIPS-enabled-Kubernetes-cluster.md

```diff
@@ -0,0 +1,134 @@
+
+### SCRAM-SHA credentials
+
+Another issue we had to deal with was the credentials for SCRAM-SHA-512 authentication.
+The SCRAM-SHA-512 authentication which is supported by Strimzi is FIPS compatible.
+But to work properly, it requires the passwords to be at least 32 characters long.
+If the password is shorter, you will get an exception similar to this:
+
+```
+Caused by: java.security.InvalidKeyException: init() failed
+        at sun.security.pkcs11.P11Mac.engineInit(P11Mac.java:228) ~[jdk.crypto.cryptoki:?]
+        at javax.crypto.Mac.chooseProvider(Mac.java:365) ~[?:?]
+        at javax.crypto.Mac.init(Mac.java:434) ~[?:?]
+        ...
+Caused by: sun.security.pkcs11.wrapper.PKCS11Exception: CKR_KEY_SIZE_RANGE
+        at sun.security.pkcs11.wrapper.PKCS11.C_SignInit(Native Method) ~[jdk.crypto.cryptoki:?]
+        at sun.security.pkcs11.P11Mac.initialize(P11Mac.java:186) ~[jdk.crypto.cryptoki:?]
+        at sun.security.pkcs11.P11Mac.engineInit(P11Mac.java:226) ~[jdk.crypto.cryptoki:?]
+        at javax.crypto.Mac.chooseProvider(Mac.java:365) ~[?:?]
+        at javax.crypto.Mac.init(Mac.java:434) ~[?:?]
+        ...
+```
+
+The length of the SCRAM-SHA password generated by the Strimzi User Operator is configurable.
+But by default, they are 12 characters long.
+That is why in Strimzi 0.33, we changed the default length to 32 characters to make the User Operator work on FIPS-enabled clusters out of the box.
+
+If you want to use a different password length, you can use the environment variable `STRIMZI_SCRAM_SHA_PASSWORD_LENGTH` to change it.
+The environment variable can be set in the `Kafka` custom resource in `.spec.entityOperator.template.userOperatorContainer.env` or directly in the `Deployment` of the standalone User Operator.
+
+## Upgrading
+
+Thanks to the improvements described above, when you deploy a new Apache Kafka cluster with Strimzi 0.33 or newer on a FIPS-enabled Kubernetes cluster, everything should work out of the box.
+But what if you are already running Strimzi and Apache Kafka on a FIPS-enabled Kubernetes cluster with the FIPS mode disabled through Strimzi's `FIPS_MODE` options?
+
+In such case, you can of course move to run under the FIPS mode as well.
+But to avoid any possible issues, you should go through the following steps.
+
+First, upgrade to Strimzi 0.33 or newer, but keep the `FIPS_MODE` environment variable still set to `disabled` to disable the FIPS mode.
+Once you are running the new Strimzi version, but with the FIPS mode still disabled, you should update the certificates and the SCRAM-SHA passwords:
+
+* If you initially deployed your cluster with a Strimzi version older than 0.30, it might still use the PKCS12 stores using old encryption and digest algorithms which will not be supported with FIPS enabled.
```

Comment:
```suggestion
* If you initially deployed your cluster with a Strimzi version older than 0.30, it might still use the PKCS12 stores using old encryption and digest algorithms which are not supported with FIPS enabled.
```